community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -815,6 +742,179 @@ public String toString()
 
     public void dispose()
     {
+        markAsShutdown();
         storageStatement.close();
     }
+
+    void markAsShutdown()
+    {
+        if ( transactionStatus.shutdown() )
+        {
+            // since transaction is marked as closed now and any new calls to close transaction
+            // are no longer possible we can release the locks now immediately
+            StatementLocks localLocks = this.statementLocks;
+            if ( localLocks != null )
+            {
+                localLocks.close();
+            }
+        }
+    }
+
+    /**
+     * It is not allowed for the same transaction to perform database writes as well as schema writes.
+     * This enum tracks the current write transactionStatus of the transaction, allowing it to transition from
+     * no writes (NONE) to data writes (DATA) or schema writes (SCHEMA), but it cannot transition between
+     * DATA and SCHEMA without throwing an InvalidTransactionTypeKernelException. Note that this behavior
+     * is orthogonal to the SecurityContext which manages what the transaction or statement is allowed to do
+     * based on authorization.
+     */
+    private enum TransactionWriteState
+    {
+        NONE,
+        DATA
+                {
+                    @Override
+                    TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKernelException
+                    {
+                        throw new InvalidTransactionTypeKernelException(
+                                "Cannot perform schema updates in a transaction that has performed data updates." );
+                    }
+                },
+        SCHEMA
+                {
+                    @Override
+                    TransactionWriteState upgradeToDataWrites() throws InvalidTransactionTypeKernelException
+                    {
+                        throw new InvalidTransactionTypeKernelException(
+                                "Cannot perform data updates in a transaction that has performed schema updates." );
+                    }
+                };
+
+        TransactionWriteState upgradeToDataWrites() throws InvalidTransactionTypeKernelException
+        {
+            return DATA;
+        }
+
+        TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKernelException
+        {
+            return SCHEMA;
+        }
+    }
+
+    static class TransactionStatus
```

Comment:
I wonder why this class (and the changes it involves) was introduced. In frequently accessed methods in KTI, instead of checking a non-volatile boolean the execution path now includes following reference to TransactionStatus object, following reference to AtomicInteger and incur volatile read/write on internal state in it. That in itself is a big difference. Why is this change made in this PR? Why is it made at all? Has his change been properly benchmarked (micro and macro)?